# Desafio Evaluado Semana 2

## Req 1: Generacion de funciones

Genere funciones para calcular la media y varianza de un vector. Debe cumplir con los siguientes requisitos:

* Ambas funciones deben ingresar un argumento x correspondienteal vector.
* Las funciones debencontenerdocstrings conladocumentaciónasociadaa la variable y retornar un valor utilizandoreturn.
* La función de la varianza se debe llamar a la función de la media.

In [1]:
import numpy as np
import pandas as pd

![Formula media](media-formula.jpg "Formula M")

In [2]:
def calc_media(x):
    """[Calcula la media de un vector]
    :param x: [Vector que contiene los valores]
    :type x: [Vector: Array]
    :return: [La media del vector]
    :rtype: [float]
    """
    sumatoria = 0
    for valor in x:
        sumatoria += valor
    return  sumatoria / len(x)

In [3]:
#validacion vs numpy
ejemplo_media = [1 , 4, 5, 6, 7]
calc_media(ejemplo_media) == np.mean(ejemplo_media)

True

![Formula Desviacion estandar](varianza_formula.jpg "Desviacion Estandar")

In [13]:
def calc_varianza(x):
    """[Calcula la varianza de un vector]
    :param x: [Vector que contiene los valores]
    :type x: [Vector: Array]
    :return: [Varianza del vector]
    :rtype: [float]
    """
    sumatoria = 0
    for valor in x:
        sumatoria += ((valor - calc_media(x))**2)        
    return sumatoria / (len(x))

In [14]:
#validacion vs numpy
calc_varianza(ejemplo_media) == np.var(ejemplo_media)

True

* Utilice las funciones para reportar la información sobre **goles_favor**, **goles_contra** y **puntos**.

In [6]:
df = pd.read_csv('worldcup2014.csv')

In [7]:
df.head()

team     continent group  group_pos  cantidad_juegos  \
0       Brazil  southamerica     A          1                3   
1       Mexico  northamerica     A          2                3   
2      Croatia        europe     A          3                3   
3     Cameroon        africa     A          4                3   
4  Netherlands        europe     B          1                3   

   juegos_ganados  juegos_empatados  juegos_perdidos  goles_favor  \
0               2                 1                0            7   
1               2                 1                0            4   
2               1                 0                2            6   
3               0                 0                3            1   
4               3                 0                0           10   

   goles_contra goles_diferencia  puntos  clasificado  
0             2                5       7            1  
1             1                3       7            1  
2             6                0       3            0  
3             9               −8       0            0  
4             3                7       9            1

In [10]:
# Funciones creadas
print(f"media goles a favor: {calc_media(df['goles_favor'])}")
print(f"media goles a contra: {calc_media(df['goles_contra'])}")
print(f"media puntos: {calc_media(df['puntos'])}")
print(f"varianza goles_favor: {calc_varianza(df['goles_favor'])}")
print(f"varianza goles_contra: {calc_varianza(df['goles_contra'])}")
print(f"varianza puntos: {calc_varianza(df['puntos'])}")

media goles a favor: 4.25
media goles a contra: 4.25
media puntos: 4.21875
varianza goles_favor: 5.1875
varianza goles_contra: 4.8125
varianza puntos: 8.0458984375


## Req 2

Utilizando el método groupby de la clase DataFrame, en conjunto con lafunción.agg de pandas, calcule la media, la varianza y desviación estándar de la cantidad de goles a favor, en contra y de la cantidad de puntos por continente.

* ¿En qué continente se observa una mayor cantidad de goles a favor?
* ¿En qué continente se observa una mayor cantidad de goles en contra?
* ¿En qué continente se observa una mayor cantidad de puntos en promedio?

In [64]:
df.groupby('continent').agg({'goles_favor':[sum], 'goles_contra':[sum],'puntos':[np.mean]})


goles_favor goles_contra    puntos
                     sum          sum      mean
continent                                      
africa                18           28  2.400000
asia                   9           25  0.750000
europe                62           52  4.692308
northamerica          13           14  4.500000
southamerica          34           17  6.833333

* Europa con 62 goles a favor sumando todos los paises europeos
* Europa con 52 goles en contra sumando todos los paises europeos
* Sudamerica 6.83 puntos en en promedio de los paises sudamericanos

## Req 3: Simulaciones

* Genere una función generate_pet que devuelva de forma aleatoria un string 'perro' o'gato' un número n de veces. Ejecútela un par de veces.

In [193]:
def generate_pet(n_veces = 2):
    """[Genera n_veces una simulacion]
    :param n_veces: [Opcional, por defecto 2. Numero de veces a generar simulacion]
    :type n_veces: [Int]
    :return: [ Data Frame contenedora de la simulacion columna pets]
    :rtype: [pandas.core.frame.DataFrame]
    """
    df = pd.DataFrame(np.random.choice(['gato', 'perro'], n_veces), columns=['Pets'])
    return df



In [194]:
generate_pet()

Pets
0  gato
1  gato

In [195]:
generate_pet(10)

Pets
0   gato
1   gato
2  perro
3  perro
4  perro
5   gato
6   gato
7   gato
8  perro
9  perro

* Aplique la función **generate_pet** para generar 20 muestras.

In [196]:
muestra_20 = generate_pet(20)
muestra_20

Pets
0   perro
1    gato
2    gato
3   perro
4    gato
5    gato
6   perro
7   perro
8   perro
9    gato
10   gato
11   gato
12   gato
13  perro
14  perro
15  perro
16  perro
17   gato
18   gato
19   gato

* ¿Cuál es la probabilidad de elegir un perro al azar? ¿Y un gato?

In [197]:
muestra_20.value_counts(normalize=True)

Pets 
gato     0.55
perro    0.45
dtype: float64

* Agregue np.random.seed(2) al inicio del chunk. ¿Qué diferencia hay cuando se ejecuta la función varias veces luego de fijar la semilla?

In [199]:
def generate_pet2(n_veces = 2):
    """[Genera n_veces una simulacion]
    :param n_veces: [Opcional, por defecto 2. Numero de veces a generar simulacion]
    :type n_veces: [Int]
    :return: [ Data Frame contenedora de la simulacion columna pets]
    :rtype: [pandas.core.frame.DataFrame]
    """
    np.random.seed(2)
    df = pd.DataFrame(np.random.choice(['gato', 'perro'], n_veces), columns=['Pets'])
    return df

In [216]:
generate_pet2(10) == generate_pet2(10)

Pets
0  True
1  True
2  True
3  True
4  True
5  True
6  True
7  True
8  True
9  True

La lista generada se mantiene fija y no cambia

## Req 4: Funcion Simuladora

* Genere una función llamada simulate_pets_prob que tome como argumento un número finito de simulaciones a generar.
* La función debe simular dos situaciones young_pet y old_pet, y contar la ocurrencia de los siguientes casos:

* De las dos mascotas simuladas (young y old), contar las ocasiones donde por lo menos una de las mascotas sea un perro
* De las dos mascotas simuladas, contar las ocasiones donde old_pet sea un perro.
* De las dos mascotas simuladas, contar las ocasiones donde los dos sean perros.

* El método debe tener una semilla pseudoaleatoria de 1.
* El output de la función debe entregarse en términos de Probabilidad.


In [302]:
def simulate_pet_prob(n_veces = 2):
    """[Genera n_veces una simulacion e imprime en pantalla las probabilidades relacionada a eventos determinados]
    :param n_veces: [Opcional, por defecto 2. Numero de veces a generar simulacion]
    :type n_veces: [Int]
    :return: []
    :rtype: []
    """
    np.random.seed(1)
    young_pet = np.random.choice(['gato', 'perro'], n_veces)
    old_pet = np.random.choice(['gato', 'perro'], n_veces)

    #al menos 1 perro
    df_old_young = pd.DataFrame({'young_pet':young_pet, 'old_pet':old_pet})
    df_old_young['almenos_un_perro'] = np.logical_or(df_old_young['young_pet']=='perro', 
                                                 df_old_young['old_pet']=='perro')
    prob_almenos_1_perro = df_old_young['almenos_un_perro'].value_counts()[True] / n_veces
    
    #solo old = perro
    old_perro = len(df_old_young[df_old_young['old_pet']=='perro']) / n_veces
    
    #ambos perro
    df_old_young['ambos_perro'] = np.logical_and(df_old_young['young_pet']=='perro', 
                                                 df_old_young['old_pet']=='perro')
    ambos_perro = df_old_young['ambos_perro'].value_counts()[True] / n_veces                     
    
    print(f'''
    Prob de: Almenos una mascota es perro {prob_almenos_1_perro*100} %
    Prob de: Solo old pet es perro {old_perro*100} %
    Prob de: Ambos son perros {ambos_perro*100} %
    ''')

In [303]:
simulate_pet_prob(10)


    Prob de: Almenos una mascota es perro 80.0 %
    Prob de: Solo old pet es perro 40.0 %
    Prob de: Ambos son perros 30.0 %
    


* De los tres escenarios, ¿Cuál es el menos probable? ¿Cuál es el más probable?¿Por qué?

El escenario menos probable es que ambos sean perros, ya que esta condicionando que se cumpla que ambos sean perros excluyentemente

El mas probable es que al menos una mascota sea perro. Esto se da por que la condicion se puede cumplir en una u otra lista, lo cual aumenta la posibilidad de que almenos uno sea un perro.